src/fun/parser.rs

```diff
@@ -1090,7 +1090,9 @@ pub trait ParserCommons<'a>: Parser<'a> {
     };
     let num_str = self.take_while(move |c| c.is_digit(radix) || c == '_');
     let num_str = num_str.chars().filter(|c| *c != '_').collect::<String>();
-    if num_str.is_empty() {
+    if let Some(c) = self.peek_one() && c.is_numeric() {
+      self.expected("valid digit")
```

Comment:
Instead of just "valid digit", would be better to show the specific radix name, so check the radix and then "valid binary digit", "valid decimal digit" or "valid hexadecimal digit"